src/Microsoft.OData.Core/UriParser/Binders/MetadataBindingUtils.cs

```diff
@@ -58,19 +58,31 @@ internal static SingleValueNode ConvertToTypeIfNeeded(SingleValueNode source, IE
                 }
 
                 ConstantNode constantNode = source as ConstantNode;
-                if (constantNode != null && constantNode.Value != null && source.TypeReference.IsString() && targetTypeReference.IsEnum())
+                // Check if the source node is a constant node, not null, and the source type is either string or integral
+                // and the target type is an enum.
+                if (constantNode != null && constantNode.Value != null && (source.TypeReference.IsString() || source.TypeReference.IsIntegral()) && targetTypeReference.IsEnum())

```

Comment:
Curious about `IsIntegral`. Does it return true for floating point numbers? If it does, that could be cause some grief. At least confirm that if someone uses a floating point number (e.g. `$filter=Gender eq 1.0`), the logic handles that gracefully and the expected exemption is thrown

Reply:
I believe Integral precludes floats, but I stand to be corrected. A confirmation with a test would great.

Reply:
IsIntegral covers: 
               case EdmPrimitiveTypeKind.Int64:
                case EdmPrimitiveTypeKind.Int32:
                case EdmPrimitiveTypeKind.Int16:
                case EdmPrimitiveTypeKind.Byte:
                case EdmPrimitiveTypeKind.SByte:

Reply:
@gathogojr, @habbes , @xuzhg

Floating point numbers will throw `CannotConvertToType` exception when used. Here is the test for this:
- [IfTypePromotionNeeded_SourceIsFloatMemberValuesAndTargetIsEnum_CannotConvertToTypeExceptionIsThrown()](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/test/FunctionalTests/Microsoft.OData.Core.Tests/UriParser/Binders/MetadataBindingUtilsTests.cs#L181)